### ISSUE TITLE
Multiple armv8 SIMD integer instructions decompiled wrong

### ISSUE THREAD

**Issue by rfalke** [NationalSecurityAgency/ghidra#3957]
**Describe the bug**
There is a test subject which is so constructed that each function tests a specific armv8 SIMD integer instruction. If the decompiler is correct it will write 3 zeros to global variables. ghidra doesn't so this in call cases.

**To Reproduce**
Decompile https://github.com/rfalke/decompiler-subjects/blob/master/from_holdec/stress_arm64/arm64_macho_simdInt_single_inst__1_var/subject.exe

**Expected behavior**
All functions are decompiled like
```
undefined8 _inst_7_var_0(void)

{
  _dest_gp = 0;
  _dest_high = 0;
  _dest_low = 0;
  return 0;
}
```

Instead sometimes ghidra doesn't have an implementation for an instruction like in:
```
undefined8 _inst_64_var_0(void)

{
  undefined auVar1 [16];
  
  auVar1 = NEON_clz(CONCAT115(0x82,CONCAT114(0x47,CONCAT113(0x4b,CONCAT112(0x48,CONCAT111(0xb8,
                                                  CONCAT110(0x58,CONCAT19(0x53,CONCAT18(0xe2,
                                                  0x8eac03f1892825da)))))))),2);
  _dest_gp = 0;
  _dest_high = (SUB168(auVar1 >> 0x40,0) ^ 0xceb48b931707638) + 0x7db8b4b647a7ac1d;
  _dest_low = (SUB168(auVar1,0) ^ 0x8eeb03b931707638) + 0x7153fc0876d7da28;
  return 0;
}
```

Or (and this is the scope of this bug report) decompiles it wrong. As in:
```
undefined8 _inst_33_var_0(void)

{
  _dest_gp = 0;
  _dest_high = 0x5e77ffff556d0000;
  _dest_low = 0;
  return 0;
}
```

**Attachments**

[ghidra_armv7_issues.txt](https://github.com/NationalSecurityAgency/ghidra/files/7983713/ghidra_armv7_issues.txt)

**Environment (please complete the following information):**
 - OS: linux
 - Ghidra Version: ghidra_10.1.2_PUBLIC

**Comment by CmP-lt** [comment]
It looks like implementations of most instructions that are listed in ghidra_armv8_issues.txt don't pass internal tests due to one or more of the following reasons (as can be checked in [AARCH64neon.sinc](https://github.com/NationalSecurityAgency/ghidra/blob/894bd3cfc2db78686727bbf2a1d369de92878033/Ghidra/Processors/AARCH64/data/languages/AARCH64neon.sinc)):
- "nofpround"
- "nointround"
- "nointsat"

@GhidorahRex are there any workarounds for instructions with semantics that can't be implemented accurately due to absence of native support for floating-point/integer rounding modes or it is more of a fundamental limitation that can't be overcome without extending sleigh?

**Comment by GhidorahRex** [comment]
@CmP-lt For the most part, the workaround is to use a pcode op. I recognize that it's going to lead to some ugly decompilation for some of those situations. Rounding modes and saturation are tricky, and sleigh doesn't really have the intrinsics to make those instructions play nicely without a lot of additional pcode - and that will most likely end up making the decompilation look just as ugly in different ways.

If something is implemented and wrong though, that's a bigger cause for concern.

**Comment by GhidorahRex** [comment]
It's a complicated issue. In one processor, we have a flag that allows the switch of emulated vs. decompiled code output. The emulated output is "technically correct" and the decompiled code is simplified. But that sort of thing doesn't necessarily work well for math-heavy applications where you're trying to represent complex arithmetic succinctly in decompiled code and very accurate.

For saturation, the workaround may be to actually implement it. I tried to do that for one of the neon instructions, `sqadd`, in the sample. It worked very well (decompile was much improved) but the answers weren't quite right because the saturation arithmetic is still funky.

**Comment by CmP-lt** [comment]
Worth to note that scope of this issue is broader than only implementations of some AARCH64 instructions, apparently there are test cases that are decompiled incorrectly despite implementations of tested instructions being [supposedly] correct.

For example, test case 33 (function `_inst_33_var_0`), despite `addp` instruction that is used in it having correct implementation, is decompiled as:
```
undefined8 _inst_33_var_0(void)

{
  _dest_high = 0x5e77ffff556d0000;
  _dest_low = 0;
  _dest_gp = 0;
  _dest_flags = 0;
  return 0;
}
```

Other example that appears in several test cases is wrong decompilation that references address in register space that is not used anywhere in function's raw pcode as happens in test case 229 (function `_inst_229_var_0`) that is decompiled as follows:
```
undefined8 _inst_229_var_0(void)

{
  ulong uVar1;
  short sVar2;
  short sVar3;
  short sVar4;
  short in_register_00005366;
  
  sVar2 = in_register_00005366 * -0x907 + -0x907;
  uVar1 = (ulong)CONCAT22(in_register_00005366 * -0x3ef9 + -0x3ef9,sVar2);
  sVar3 = in_register_00005366 * 0x29dd + 0x29dd;
  sVar4 = in_register_00005366 * 0xbdf + 0xbdf;
  _dest_high = 0;
  _dest_low = (CONCAT17((char)((ushort)sVar4 >> 8),
                        CONCAT16((char)sVar4,
                                 CONCAT15((char)((ushort)sVar3 >> 8),
                                          CONCAT14((char)sVar3,
                                                   CONCAT13((char)(uVar1 >> 0x18),
                                                            CONCAT12((char)(uVar1 >> 0x10),sVar2)) ))
                                )) ^ 0x6d5b8d83119e1a6b) + 0xa5805610b7e523e9;
  _dest_gp = 0;
  _dest_flags = 0;
  return 0;
}
```

So I think that discussing all causes of incorrect decompilation of test cases from provided test subject in one issue isn't the best option, instead it makes more sense to create separate issue for each identified cause or group of causes. I plan to do that at least for 2 causes illustrated above, but will first need to do some research on them (in hope to get any insights that might be helpful for diagnosing the issues).

**Comment by CmP-lt** [comment]
> The register address is used by the raw pcode, but it's incorrect.

You are right, this one is an oversight from my side, thanks for correction. So this cause is also something to be fixed in sleigh. For completeness here is the list of test cases that have wrong decompilation due to this cause:
- `_inst_229_var_0`
- `_inst_231_var_0`
- `_inst_241_var_0`
- `_inst_280_var_0`
- `_inst_282_var_0`
- `_inst_521_var_0`
- `_inst_525_var_0`
- `_inst_531_var_0`
- `_inst_540_var_0`
- `_inst_886_var_0`
- `_inst_891_var_0`
- `_inst_901_var_0`
- `_inst_910_var_0`

**Comment by CmP-lt** [comment]
It appears that several cases, including `_inst_33_var_0`, have been fixed by changes in decompiler since 10.3 (most likely in 8977840661f66050055bf17985864fb5973d07ff). Fixed cases include: 33, 248, 788, 830, 1054, 1058, 1081, 1085, 1087.

But it's also worth to add that case `_inst_27_var_0` that also tests `addp` instruction (but with different operand types) decompiles not as expected both in 10.3 and with latest changes to decompiler. It's decompilation looks correct, but something prevents complete simplification of upper 8 bytes:
```
undefined8 _inst_27_var_0(void)

{
  undefined auVar1 [16];
  
  auVar1[8] = 0xac;
  auVar1._0_8_ = 0x47ba0b5ab88a1aa1;
  auVar1[9] = 0xc1;
  auVar1[10] = 0xb;
  auVar1[11] = 0x81;
  auVar1[12] = 0x3e;
  auVar1[13] = 0x62;
  auVar1[14] = 0x68;
  auVar1[15] = 0xee;
  _dest_high = auVar1._8_8_ + 0x11979dc17ef43e54;
  _dest_low = 0;
  _dest_gp = 0;
  _dest_flags = 0;
  return 0;
}
```

**Comment by rfalke** [comment]
For ghidra_10.1.2_PUBLIC the summary is:

```
correct (with the expected zeros):         422
incorrect (with some constant):            43
incorrect (with some complex computation): 623
```

For ghidra_10.3.2_PUBLIC the summary is:
```
correct (with the expected zeros):         445
incorrect (with some constant):            44
incorrect (with some complex computation): 599
```

The new file [ghidra_armv8_issues_ghidra_10.3.2_PUBLIC.txt](https://github.com/NationalSecurityAgency/ghidra/files/12423177/ghidra_armv8_issues_ghidra_10.3.2_PUBLIC.txt) has the information about the 44 cases where the computation is wrong but not so complex.

**Comment by CmP-lt** [comment]
Most of remaining cases from group "incorrect (with some constant)" can at best be partially fixed to fall into category of functions that are decompiled with computations by adding various pcodeops to represent rounding modes. But even such partial measure should be better than nothing, because currently those instructions simply have missing operations (i.e. rounding) in their implementations.
